### PR TITLE
Reactions dev

### PR DESCRIPTION
**Pull request by swetha-gangu**

### CODE REVIEW

File: src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -1,8 +1,10 @@
 package com.google.sps.data;
 
 import java.util.Objects;
+import com.google.sps.data.Constants;
 
 public final class Comment{
+    private static final double COMMENT_TOXICITY_THRESHOLD = 0.8;
```
Comment (jindaln) [inline src/main/java/com/google/sps/data/Comment.java:7]: why 0.8 - maybe add a comment explaining the threshold choice?

Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:7]: this PR is for reactions right? why is there toxicity stuff in here at all?

File: src/main/java/com/google/sps/data/Constants.java [hunk 2]
```diff
@@ -6,6 +6,9 @@
     public static final String REP_TITLE = "Official Title";
     public static final String REP_POSTS = "Posts";
 
+    public static final String PERSPECTIVE_API_KEY = "PERSPECTIVE_API_KEY";
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Constants.java:9]: same for these

File: src/main/java/com/google/sps/data/Reaction.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package com.google.sps.data;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** 
+ * Reactions a user can add to a post 
+ */
+public enum Reaction {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Reaction.java:10]: where is toString defined for this?

File: src/main/java/com/google/sps/data/Reaction.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package com.google.sps.data;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** 
+ * Reactions a user can add to a post 
+ */
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Reaction.java:9]: see a better example of an enum

https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html

File: src/test/java/com/google/sps/data/CommentTest.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import com.google.sps.data.Comment;
+import com.google.sps.data.Representative;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class CommentTest {
```
Comment (v1shanker) [inline src/test/java/com/google/sps/data/CommentTest.java:13]: you probably don't need this test. it does nothing. also see other comment about why toxicity is in the PR.

File: src/main/java/com/google/sps/data/Comment.java (hunk 1 above)
Comment (swetha-gangu) [reply src/main/java/com/google/sps/data/Comment.java:7]: it's just something random for now and we can decide later

Comment (swetha-gangu) [reply src/main/java/com/google/sps/data/Comment.java:7]: I just have it in here from when I was working on perspective. I'll go ahead and delete

File: src/main/java/com/google/sps/data/Reaction.java (hunk 3 above)
Comment (swetha-gangu) [reply src/main/java/com/google/sps/data/Reaction.java:10]: the default toString() was sufficient

File: src/test/java/com/google/sps/data/CommentTest.java (hunk 5 above)
Comment (swetha-gangu) [reply src/test/java/com/google/sps/data/CommentTest.java:13]: I'll remove

File: src/main/java/com/google/sps/data/Comment.java [hunk 6]
```diff
@@ -1,8 +1,9 @@
 package com.google.sps.data;
 
 import java.util.Objects;
+import com.google.sps.data.Constants;
 
-public final class Comment{
+public final class Comment{
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:6]: remote extra space at end of line and add space between Comment and {

File: src/main/java/com/google/sps/data/Comment.java [hunk 7]
```diff
@@ -12,7 +13,7 @@ public Comment(String name, String comment, long id){
         this.comment = comment;
         this.id = id; 
     }
-
+
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:16]: newline is good but it doesn't need spaces

File: src/main/java/com/google/sps/data/DatastoreManager.java [hunk 8]
```diff
@@ -198,4 +203,47 @@ public static Entity queryForPostEntityWithId(long postId)
...
+        postEntity.setProperty(reaction, reactionCount + 1); 
+        datastore.put(postEntity);
+    }
+
+    /**
+     * Removes a reaction from a post entity 
+     * @param postId ID of the post entity 
+     * @param reaction String of the reaction enum 
+     */ 
+    public static void removeReactionFromPost(long postId, String reaction)
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreManager.java:232]: should removing a reaction from a post that doesn't already have one be an error?

File: src/main/java/com/google/sps/data/DatastoreManager.java [hunk 9]
```diff
@@ -198,4 +203,47 @@ public static Entity queryForPostEntityWithId(long postId)
...
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity postEntity = DatastoreManager.queryForPostEntityWithId(postId); 
+        long reactionCount; 
+        if (postEntity.getProperty(reaction) != null) {
+            reactionCount = (long) postEntity.getProperty(reaction); 
+        }
+        else {
+            reactionCount = 0; 
+        }
+        if (reactionCount != 0) {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreManager.java:243]: you can move the - 1 into the if postEntity.getProperty(reaction) != null clause.

File: src/main/java/com/google/sps/data/Post.java [hunk 10]
```diff
@@ -1,21 +1,35 @@
...
         this.question = question;
         this.answer = answer;
         this.replies = replies;
         this.id = id;
+        this.reactions = reactions; 
+    }
+
+    public Map<Reaction, Long> getReactions() {
+        Map<Reaction, Long> copy = new HashMap<Reaction, Long>();
+        for (Map.Entry<Reaction, Long> entry : reactions.entrySet())
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Post.java:28]: I believe Hashmap has a constructor that can take another map - see if that helps.

File: src/main/java/com/google/sps/data/Reaction.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+public enum Reaction {
+    THUMBS_UP; 
+
+    public static Reaction stringToEnumReaction(String reactionStr) {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Reaction.java:13]: call this method fromString

File: src/main/java/com/google/sps/data/Reaction.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        return Reaction.valueOf(reactionStr); 
+    }
+
+    public static List<String> getReactionsAsStrings() {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Reaction.java:17]: allValues()

since it will be Reactions.allValues() it reads cleaner

File: src/main/java/com/google/sps/data/Reaction.java (hunk 3 above)
Comment (v1shanker) [reply src/main/java/com/google/sps/data/Reaction.java:10]: please define a string value for THUMBS_UP:

THUMBS_UP ("thumbs_up");

I understand the default is sufficient but please define it anyways so it doesn't keep the reader guessing. Especially since the value is important in the datastore and the front end.

File: src/test/java/com/google/sps/servlets/UnreactToPostServletTest.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    public void testDoPost() throws Exception {
+        // add Representative entity to Datastore 
+        long questionId = DatastoreManager.insertCommentInDatastore("Bob", "Why are you president?"); 
+        Long postId = DatastoreManager.insertPostInDatastore(questionId); 
+        when(request.getParameter("postId")).thenReturn(postId.toString());
+        when(request.getParameter("reaction")).thenReturn(Reaction.THUMBS_UP.toString());
+        when(request.getParameter("repName")).thenReturn("Donald Trump");
+        long repId = DatastoreManager.insertRepresentativeInDatastore("Donald Trump", "President");
+        DatastoreManager.updateRepresentativePostList(repId, postId); 
+        DatastoreManager.addReactionToPost(postId, Reaction.THUMBS_UP.toString());
```
Comment (jindaln) [inline src/test/java/com/google/sps/servlets/UnreactToPostServletTest.java:63]: Just a clarification - why is addReaction happening twice?

File: src/test/java/com/google/sps/servlets/UnreactToPostServletTest.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        Map<Reaction, Long> reactions = new HashMap<Reaction, Long>(); 
+        reactions.put(Reaction.THUMBS_UP, (long) 1);
+        Post post = new Post(question, null, replies, postId, reactions); 
+        posts.add(post); 
+        Representative expectedRep = new Representative("Donald Trump", "President", posts, repId);
+
+        servlet.doGet(request, response);
+    
+        Representative actualRep = DatastoreManager.queryForRepresentativeObjectWithName("Donald Trump");
+        assertTrue(actualRep.equals(expectedRep));
```
Comment (jindaln) [inline src/test/java/com/google/sps/servlets/UnreactToPostServletTest.java:79]: Is the point of the test to check if the representative's are the same or just to check that the number of reactions is what you expect it to be? As in, is it necessary to create the whole rep object or can you just assert # of reactions

File: src/test/java/com/google/sps/servlets/ReactToPostServletTest.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        Map<Reaction, Long> reactions = new HashMap<Reaction, Long>(); 
+        reactions.put(Reaction.THUMBS_UP, (long) 1);
+        Post post = new Post(question, null, replies, postId, reactions); 
+        posts.add(post); 
+        Representative expectedRep = new Representative("Donald Trump", "President", posts, repId);
+
+        servlet.doGet(request, response);
+    
+        Representative actualRep = DatastoreManager.queryForRepresentativeObjectWithName("Donald Trump");
+        assertTrue(actualRep.equals(expectedRep));
```
Comment (jindaln) [inline src/test/java/com/google/sps/servlets/ReactToPostServletTest.java:77]: Same comment as for unreact test - do you need to assert the whole rep?

File: src/test/java/com/google/sps/data/DatastoreManagerTest.java [hunk 16]
```diff
@@ -233,4 +233,63 @@ public void testQueryForPostEntityWithId()
...
+
+        DatastoreManager.removeReactionFromPost(postId, Reaction.THUMBS_UP.toString()); 
+        Post post = DatastoreManager.queryForPostObjectWithId(postId); 
+        long reactionCount = post.getReactions().get(Reaction.THUMBS_UP); 
+
+        assertTrue(reactionCount == 0);
+    }
+
+    @Test
+    public void testRemoveReactionFromPostWithTwoReactions()
```
Comment (jindaln) [inline src/test/java/com/google/sps/data/DatastoreManagerTest.java:281]: Is there an edge case here that you wouldn't cover with testRemoveReactionFromPostWithOneReaction()

File: src/main/java/com/google/sps/data/DatastoreManager.java (hunk 9 above)
Comment (swetha-gangu) [reply src/main/java/com/google/sps/data/DatastoreManager.java:243]: what if reaction isn't null but it's set to 0?

File: src/main/java/com/google/sps/data/DatastoreManager.java (hunk 8 above)
Comment (swetha-gangu) [reply src/main/java/com/google/sps/data/DatastoreManager.java:232]: yes

File: src/main/java/com/google/sps/data/DatastoreManager.java (hunk 9 above)
Comment (swetha-gangu) [reply src/main/java/com/google/sps/data/DatastoreManager.java:243]: ill just account for that and move it into the !=null block

File: src/test/java/com/google/sps/data/DatastoreManagerTest.java (hunk 16 above)
Comment (swetha-gangu) [reply src/test/java/com/google/sps/data/DatastoreManagerTest.java:281]: I can't remember why I have this. I'll go ahead and remove.

File: src/main/java/com/google/sps/data/DatastoreManager.java [hunk 17]
```diff
@@ -198,4 +204,44 @@ public static Entity queryForPostEntityWithId(long postId)
...
+    /**
+     * Removes a reaction from a post entity 
+     * @param postId ID of the post entity 
+     * @param reaction String of the reaction enum 
+     */ 
+    public static void removeReactionFromPost(long postId, String reaction) 
+    throws EntityNotFoundException, UnsupportedOperationException {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity postEntity = DatastoreManager.queryForPostEntityWithId(postId); 
+        if (postEntity.getProperty(reaction) == null || (long) postEntity.getProperty(reaction) == 0) {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreManager.java:237]: break line since its > 100 chars

File: src/main/java/com/google/sps/data/DatastoreManager.java [hunk 18]
```diff
@@ -198,4 +204,44 @@ public static Entity queryForPostEntityWithId(long postId)
...
+            throw new UnsupportedOperationException("No such reactions in post entity"); 
+        }
+        else {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreManager.java:240]: since the above line returns early, the rest of the function doesn't need to be in a else block

less indentation is more readable

File: src/main/java/com/google/sps/data/DatastoreManager.java (hunk 17 above)
Comment (v1shanker) [inline src/main/java/com/google/sps/data/DatastoreManager.java:237]: save postEntity.getPropertly in the above line since its reused 3 times

File: src/main/java/com/google/sps/data/Reaction.java [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+    public static List<String> allValues() {
+        List<Reaction> reactionEnums = Arrays.asList(Reaction.values());
+        List<String> reactionStrings = new ArrayList<String>(); 
+        for (Reaction reaction : reactionEnums) {
+            reactionStrings.add(reaction.toString()); 
+        }
+        return reactionStrings; 
+    }
+
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Reaction.java:26]: dead code?

File: src/main/java/com/google/sps/servlets/ReactToPostServlet.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+        String reaction = request.getParameter(REACTION);
+       
+        try {
+            DatastoreManager.addReactionToPost(postId, reaction); 
+        } 
+        catch(EntityNotFoundException e) {
+            logger.error(e);
+            throw new ServletException("Error: " + e.getMessage(), e);
+        }
+        String redirect = "feed.html?name=" + repName;
```
Comment (jindaln) [inline src/main/java/com/google/sps/servlets/ReactToPostServlet.java:50]: Make sure to urlEncode this

File: src/main/java/com/google/sps/servlets/UnreactToPostServlet.java [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+        } 
+        catch(EntityNotFoundException e) {
+            logger.error(e);
+            throw new ServletException("Error: " + e.getMessage(), e);
+        }
+        catch(UnsupportedOperationException e) {
+            logger.error(e);
+            throw new ServletException("Error: " + e.getMessage(), e);
+        }
+        String redirect = "feed.html?name=" + repName;
```
Comment (jindaln) [inline src/main/java/com/google/sps/servlets/UnreactToPostServlet.java:55]: Make sure to urlEncode repName
